flink-ml-lib/src/main/java/org/apache/flink/ml/common/gbt/GBTRunner.java

```diff
@@ -0,0 +1,247 @@
+    static DataStream<GBTModelData> train(Table dataTable, GbtParams p) {
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) dataTable).getTableEnvironment();
+        Tuple2<Table, DataStream<FeatureMeta>> preprocessResult =
+                p.isInputVector
+                        ? Preprocess.preprocessVecCol(dataTable, p)
+                        : Preprocess.preprocessCols(dataTable, p);
+        dataTable = preprocessResult.f0;
+        DataStream<FeatureMeta> featureMeta = preprocessResult.f1;
+
+        DataStream<Row> data = tEnv.toDataStream(dataTable);
+        DataStream<Tuple2<Double, Long>> labelSumCount =
+                DataStreamUtils.aggregate(data, new LabelSumCountFunction(p.labelCol));
+        return boost(dataTable, p, featureMeta, labelSumCount);
+    }
+
+    private static DataStream<GBTModelData> boost(
+            Table dataTable,
+            GbtParams p,
+            DataStream<FeatureMeta> featureMeta,
+            DataStream<Tuple2<Double, Long>> labelSumCount) {
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) dataTable).getTableEnvironment();
+
+        final String featureMetaBcName = "featureMeta";
+        final String labelSumCountBcName = "labelSumCount";
+        Map<String, DataStream<?>> bcMap = new HashMap<>();
+        bcMap.put(featureMetaBcName, featureMeta);
+        bcMap.put(labelSumCountBcName, labelSumCount);
+
+        DataStream<TrainContext> initTrainContext =
+                BroadcastUtils.withBroadcastStream(
+                        Collections.singletonList(
+                                tEnv.toDataStream(tEnv.fromValues(0), Integer.class)),
+                        bcMap,
+                        (inputs) -> {
+                            //noinspection unchecked
+                            DataStream<Integer> input = (DataStream<Integer>) (inputs.get(0));
+                            return input.map(
+                                    new InitTrainContextFunction(
+                                            featureMetaBcName, labelSumCountBcName, p));
+                        });
+
+        DataStream<Row> data = tEnv.toDataStream(dataTable);
+        final IterationID iterationID = new IterationID();
+        DataStreamList dataStreamList =
+                Iterations.iterateBoundedStreamsUntilTermination(
+                        DataStreamList.of(initTrainContext.broadcast()),
+                        ReplayableDataStreamList.notReplay(data, featureMeta),
+                        IterationConfig.newBuilder()
+                                .setOperatorLifeCycle(IterationConfig.OperatorLifeCycle.ALL_ROUND)
+                                .build(),
+                        new BoostIterationBody(iterationID, p));
+        return dataStreamList.get(0);
+    }
+
+    public static GbtParams fromEstimator(BaseGBTParams<?> estimator, TaskType taskType) {
```

Comment:
All information specified in `GbtParams` can obtained from `estimator` itself. Note that all existing Estimators get the parameters directly instead of having to pass them via a container (e.g. GbtParams).

Would it be simpler and more consistent with the existing code to follow the existing practice? Spark ML does not seem to need this container for parameters. Alternative, can you explain why GBT needs to follow a different approach?

Reply:
SparkML's implementation has a class named `BoostingStrategy` [1] serving as a container like `GbtParams`.  I think its mainly purpose is to make training functions in `GBTRunner` able to be tested separately.

[1] https://github.com/apache/spark/blob/27ed89b7be5ebb91e4a0b106b1669a7867a6012d/mllib/src/main/scala/org/apache/spark/ml/classification/GBTClassifier.scala#L195